Gems/Atom/Tools/AtomToolsFramework/Code/Source/Document/AtomToolsDocument.cpp

```diff
@@ -154,4 +261,107 @@ namespace AtomToolsFramework
+        m_undoHistoryIndex = {};
+    }
+
+    bool AtomToolsDocument::OpenSucceeded()
+    {
+        AZ_TracePrintf("AtomToolsDocument", "Document opened: '%s'.\n", m_absolutePath.c_str());
+        AzToolsFramework::AssetSystemBus::Handler::BusConnect();
+        AtomToolsFramework::AtomToolsDocumentNotificationBus::Broadcast(
+            &AtomToolsFramework::AtomToolsDocumentNotificationBus::Events::OnDocumentOpened, m_id);
+        return true;
+    }
+
+    bool AtomToolsDocument::OpenFailed()
+    {
+        AZ_TracePrintf("AtomToolsDocument", "Document could not opened: '%s'.\n", m_absolutePath.c_str());
+        Clear();
+        return false;
+    }
+
+    bool AtomToolsDocument::SaveSucceeded()
+    {
+        m_ignoreSourceFileChangeToSelf = true;
+
+        AZ_TracePrintf("AtomToolsDocument", "Document saved: '%s'.\n", m_savePathNormalized.c_str());
+
+        // Auto add or checkout saved file
+        AzToolsFramework::SourceControlCommandBus::Broadcast(
+            &AzToolsFramework::SourceControlCommandBus::Events::RequestEdit, m_savePathNormalized.c_str(), true,
+            [](bool, const AzToolsFramework::SourceControlFileInfo&) {});
+
+        AtomToolsFramework::AtomToolsDocumentNotificationBus::Broadcast(
+            &AtomToolsFramework::AtomToolsDocumentNotificationBus::Events::OnDocumentSaved, m_id);
+        return true;
+    }
+
+    bool AtomToolsDocument::SaveFailed()
+    {
+        AZ_TracePrintf("AtomToolsDocument", "Document not saved: '%s'.\n", m_savePathNormalized.c_str());
+        return false;
+    }
+
+    bool AtomToolsDocument::ReopenRecordState()
+    {
+        m_undoHistoryBeforeReopen = m_undoHistory;
+        m_undoHistoryIndexBeforeReopen = m_undoHistoryIndex;
+        return true;
+    }
+
+    bool AtomToolsDocument::ReopenRestoreState()
+    {
+        m_undoHistory = m_undoHistoryBeforeReopen;
+        m_undoHistoryIndex = m_undoHistoryIndexBeforeReopen;
+        m_undoHistoryBeforeReopen = {};
+        m_undoHistoryIndexBeforeReopen = {};
+        return true;
+    }
+
+    void AtomToolsDocument::AddUndoRedoHistory(const UndoRedoFunction& undoCommand, const UndoRedoFunction& redoCommand)
+    {
+        // Wipe any state beyond the current history index
+        m_undoHistory.erase(m_undoHistory.begin() + m_undoHistoryIndex, m_undoHistory.end());
+
+        // Add undo and redo operations using functions that capture state and restore it when executed
+        m_undoHistory.emplace_back(undoCommand, redoCommand);
+
+        // Assign the index to the end of history
+        m_undoHistoryIndex = aznumeric_cast<int>(m_undoHistory.size());
+        AtomToolsFramework::AtomToolsDocumentNotificationBus::Broadcast(
+            &AtomToolsFramework::AtomToolsDocumentNotificationBus::Events::OnDocumentUndoStateChanged, m_id);
+    }
+
+    void AtomToolsDocument::SourceFileChanged(AZStd::string relativePath, AZStd::string scanFolder, [[maybe_unused]] AZ::Uuid sourceUUID)
```

Comment:
nit: const AZStd::string for relativePath and scanFolder

Reply:
do you need to add '&' as well?

Reply:
That's an override from an existing bus interface that should be updated separately to follow those good recommendations.